### PR TITLE
Reduce probability of startup port collisions

### PR DESCRIPTION
**Pull request by mikeharder**
- `GetTickCount()` is limited to the resolution of the system timer, which is typically 10-16ms.  If two apps in separate app pools are started within this time window, it's possible `GetTickCount()` will return the same value, which causes the apps to try the same random port(s).
- Addresses #1124

### CODE REVIEW

**Comment by pakrym** [comment]
Did you have a repro? Were you able to verify that conflict happens less often?

**Comment by davidfowl** [comment]
Don't we want to bind(0)? And pass that port to kestrel?

**Comment by mikeharder** [comment]
@pakrym: I can repro in both v1 and v2 outproc.  See https://github.com/aspnet/IISIntegration/issues/1124 for repro steps and log messages.  You can also repro in containers:

* https://github.com/mikeharder/ancm-docker/tree/master/app/2.1/outproc-multi-apppool
* https://github.com/mikeharder/ancm-docker/tree/master/app/2.2/outproc-multi-apppool.

In v1 outproc, if two apps select the same port, it appears the second app fails to start and never retries.

In v2 outproc, if two apps initially select the same port, it appears the second app usually retries and then does start.  However, one time I was able to repro with v2 outproc where one of the apps failed even after retry.

After this change, I was never able to repro a port collision with 10 apps in separate app pools.

**Comment by jkotalik** [comment]
I hesitate not having some sort of automated test. 

> Don't we want to bind(0)? And pass that port to kestrel?

I'd don't see how binding to 0, unbinding the port, and then passing it to kestrel is better than this? Wouldn't there be issues with kestrel thinking the port is in use? 
If we want anyone to call bind(0), its kestrel. 

@mikeharder I think there are still issues we need to address with retrying to start kestrel. There is a new loop in ANCM V2 that retries starting the process: https://github.com/aspnet/IISIntegration/blob/release/2.2/src/AspNetCoreModuleV2/OutOfProcessRequestHandler/serverprocess.cpp#L768. I think we need something equivalent in V1.

**Comment by davidfowl** [comment]
> I'd don't see how binding to 0, unbinding the port, and then passing it to kestrel is better than this? Wouldn't there be issues with kestrel thinking the port is in use?
If we want anyone to call bind(0), its kestrel.

The OS selects the "random" port in the relevant port range instead of us having to rely on the hard coded random port range specified here.

**Comment by mikeharder** [comment]
@davidfowl, @jkotalik: In my estimation, "bind to 0 and pass to kestrel" and "random" have approximately the same probability of collision (say 0.001%).  Very low, but high enough to require retry logic in either case.  So I don't think the added complexity of "bind to 0 and pass to kestrel" is worth it.  I would prefer to keep our changes to outproc (especially v1) as small as possible.

**Comment by mikeharder** [comment]
@jkotalik: In theory, we could add an automated test which does the same thing as my manual repro:

* Add 10 apps in separate app pools
* Make a request to each app in parallel (so they start as close in time as possible)
* Verify all responses are HTTP 200

With the previous code, this test would usually fail on my dev machine (at least for ANCM v1).  It should never fail with the new code.

Do you think such a test would be reliable enough and worth adding?

**Comment by pakrym** [comment]
@mikeharder I think it would be beneficial to have it both for this fix and as a startup stress scenario in general.

**Comment by jkotalik** [comment]
> Do you think such a test would be reliable enough and worth adding?
Yeah, I think so.

**Comment by mikeharder** [comment]
> > I'd don't see how binding to 0, unbinding the port, and then passing it to kestrel is better than this? Wouldn't there be issues with kestrel thinking the port is in use?
If we want anyone to call bind(0), its kestrel.

> The OS selects the "random" port in the relevant port range instead of us having to rely on the hard coded random port range specified here.

This would also be somewhat of a breaking change for ANCM v1 (different range used to select random ports).

**Comment by mikeharder** [comment]
> I think there are still issues we need to address with retrying to start kestrel. There is a new loop in ANCM V2 that retries starting the process: https://github.com/aspnet/IISIntegration/blob/release/2.2/src/AspNetCoreModuleV2/OutOfProcessRequestHandler/serverprocess.cpp#L768. I think we need something equivalent in V1.

I think this should be treated as a separate issue: https://github.com/aspnet/IISIntegration/issues/1137

**Comment by jkotalik** [comment]
As long as it is addressed 😄

**Comment by mikeharder** [comment]
@jkotalik, @pakrym: I will work on adding that test.  If we want to test this scenario thoroughly (including retries), we would need a way to use a fixed random seed for testing to force port collisions (rather than `random_device`).  In managed code this would typically be done via dependency injection (constructor or otherwise).  Is there a similar pattern for native code?  Have we done anything like this in existing tests?

**Comment by jkotalik** [comment]
>  dependency injection 

Nope 😢 

I don't think it's required to test the negative scenario of forced port collisions.

**Comment by pakrym** [comment]
@mikeharder I would not be too worried about retries. Let's just add a test that starts 10 sites all at once and makes sure they all start. If this scenario failed before it would be a good place to start.

**Comment by mikeharder** [comment]
> I don't think it's required to test the negative scenario of forced port collisions.

I agree it's not required for this PR.  But I do think we'll need a way to force collisions if we want to verify retries work correctly.

File: test/Common.FunctionalTests/Utilities/Helpers.cs [hunk 1]
```diff
@@ -25,7 +25,23 @@ public static string GetTestWebSitePath(string name)
         public static string GetInProcessTestSitesPath() => GetTestWebSitePath("InProcessWebSite");
 
         public static string GetOutOfProcessTestSitesPath() => GetTestWebSitePath("OutOfProcessWebSite");
-  
+
+        public static IISDeploymentParameters GetBaseDeploymentParameters(string site = null, HostingModel hostingModel = HostingModel.InProcess, bool publish = false)
```
Comment (pakrym) [inline test/Common.FunctionalTests/Utilities/Helpers.cs]: Why is this back?

File: test/IISExpress.FunctionalTests/OutOfProcess/MutlipleAppTests.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+using Microsoft.AspNetCore.Testing.xunit;
+using Microsoft.Extensions.Logging.Abstractions;
+using Xunit;
+
+namespace IISExpress.FunctionalTests.OutOfProcess
+{
+    public class MutlipleAppTests
+    {
+        [ConditionalFact]
+        public async Task Test()
```
Comment (pakrym) [inline test/IISExpress.FunctionalTests/OutOfProcess/MutlipleAppTests.cs]: Is there a reason this test is IISExpress specific?

File: test/IISExpress.FunctionalTests/OutOfProcess/MutlipleAppTests.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                var deployer = new IISExpressDeployer(deploymentParameters, new NullLoggerFactory());
+                deployers.Add((deployer));
+                deploymentResults[i] = await deployer.DeployAsync();
+            }
+
+            var taskList = new Task<HttpResponseMessage>[numSites];
+
+            for (var i = 0; i < numSites; i++)
+            {
+                taskList[i] = deploymentResults[i].HttpClient.GetAsync("/HelloWorld");
```
Comment (pakrym) [inline test/IISExpress.FunctionalTests/OutOfProcess/MutlipleAppTests.cs]: Can we use the stress test helper in some way?

File: test/IISExpress.FunctionalTests/OutOfProcess/MutlipleAppTests.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            const int numSites = 10;
+            var deployers = new List<IISExpressDeployer>();
+            var deploymentResults = new DeploymentResult[numSites];
+            for (var i = 0; i < numSites; i++)
+            {
+                var deploymentParameters = Helpers.GetBaseDeploymentParameters(site: "OutOfProcessWebsite");
+                deploymentParameters.AncmVersion = AncmVersion.AspNetCoreModule;
+
+                var deployer = new IISExpressDeployer(deploymentParameters, new NullLoggerFactory());
```
Comment (pakrym) [inline test/IISExpress.FunctionalTests/OutOfProcess/MutlipleAppTests.cs]: Why null logger factory?

File: test/IISExpress.FunctionalTests/OutOfProcess/MutlipleAppTests.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        [ConditionalFact]
+        public async Task Test()
+        {
+            const int numSites = 10;
+            var deployers = new List<IISExpressDeployer>();
+            var deploymentResults = new DeploymentResult[numSites];
+            for (var i = 0; i < numSites; i++)
+            {
+                var deploymentParameters = Helpers.GetBaseDeploymentParameters(site: "OutOfProcessWebsite");
+                deploymentParameters.AncmVersion = AncmVersion.AspNetCoreModule;
```
Comment (pakrym) [inline test/IISExpress.FunctionalTests/OutOfProcess/MutlipleAppTests.cs]: We need both V1 and V2

**Comment by pakrym** [comment]
Would be nice to have a test that starts 10 web applications inside a single iisexpress/w3wp instance.

**Comment by mikeharder** [comment]
@pakrym: The test is still a WIP.  I will be adding coverage for both v1 and v2, IIS and IIS Express.  It's a good idea to also test 10 applications in a single app pool for completeness, though this configuration should not be impacted by the random seed issue, since apps within an app pool are started sequentially rather than in parallel.

**Comment by jkotalik** [comment]
Sorry, I should have commented on this thread that @mikeharder and I worked on a general outline for what tests will look like.

**Comment by mikeharder** [comment]
@jkotalik, @pakrym: Ready for next round of reviews

File: test/Common.Tests/Utilities/DisposableEnumerable.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+            _enumerable = enumerable;
+        }
+
+        public void Dispose()
+        {
+            foreach (var item in _enumerable)
+            {
+                if (item != null)
+                {
+                    item.Dispose();
```
Comment (davidfowl) [inline test/Common.Tests/Utilities/DisposableEnumerable.cs]: nit: item?.Dispose()

File: test/Common.Tests/Utilities/DisposableEnumerable.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Server.IntegrationTesting
+{
+    public class DisposableEnumerable : IDisposable
```
Comment (pakrym) [inline test/Common.Tests/Utilities/DisposableEnumerable.cs]: I would've done disposable list but it doesn't matter in context of this PR.

File: test/Common.Tests/Utilities/DisposableEnumerable.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
+using System;
```
Comment (pakrym) [inline test/Common.Tests/Utilities/DisposableEnumerable.cs]: (C)

File: test/IISExpress.FunctionalTests/OutOfProcess/MultipleAppTests.cs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.IIS.FunctionalTests.Utilities;
+using Microsoft.AspNetCore.Server.IntegrationTesting;
+using Microsoft.AspNetCore.Testing.xunit;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests
+{
+    [Collection(PublishedSitesCollection.Name)]
+    public class MultipleAppTests : IISFunctionalTestBase
```
Comment (jkotalik) [inline test/IISExpress.FunctionalTests/OutOfProcess/MultipleAppTests.cs]: If this test does fail on CI, #1137 will probably be the fix.

**Pull request merged**